server/src/utils/win-conditions.ts

```diff
@@ -18,40 +42,37 @@ export const getGameOutcome = (game: GameModel, endResult: any) => {
 
 export const getGamePlayerOutcome = (
 	game: GameModel,
-	endResult: any,
-	playerId: PlayerId,
+	endResult: EndResult,
+	viewer: ViewerComponent,
 ): GamePlayerEndOutcomeT => {
 	if (Object.hasOwn(endResult, 'timeout')) return 'timeout'
 	if (Object.hasOwn(endResult, 'forfeit')) {
-		const triggerPlayerId = endResult.forfeit.playerId
-		return triggerPlayerId === playerId ? 'forfeit_loss' : 'forfeit_win'
+		const triggerPlayerId = endResult.forfeit!.playerId
+		return triggerPlayerId === getPlayerIdFromViewer(game, viewer)
+			? 'forfeit_loss'
+			: 'forfeit_win'
 	}
 	if (Object.hasOwn(endResult, 'playerRemoved')) {
-		const triggerPlayerId = endResult.playerRemoved.player.id
-		return triggerPlayerId === playerId ? 'leave_loss' : 'leave_win'
+		const triggerPlayerId = endResult.playerRemoved!.player.id
+		return triggerPlayerId === getPlayerIdFromViewer(game, viewer)
+			? 'leave_loss'
+			: 'leave_win'
 	}
 	if (game.endInfo.deadPlayerEntities.length === 2) return 'tie'
 	const deadPlayerEntity = game.endInfo.deadPlayerEntities[0]
 	if (!deadPlayerEntity) return 'unknown'
 
-	let deadPlayer = game.components.find(
-		ViewerComponent,
-		(_game, viewer) =>
-			!viewer.spectator && viewer.playerOnLeft.entity === deadPlayerEntity,
-	)
-	if (!deadPlayer) return 'unknown'
-
-	if (deadPlayer.playerId === playerId) return 'you_lost'
+	if (deadPlayerEntity === viewer.playerOnLeftEntity) return 'you_lost'
 	return 'you_won'
 }
 
-export const getWinner = (game: GameModel, endResult: any) => {
+export const getWinner = (game: GameModel, endResult: EndResult) => {
 	if (Object.hasOwn(endResult, 'timeout')) return null
 	if (Object.hasOwn(endResult, 'forfeit')) {
-		return getOpponentId(game, endResult.forfeit.playerId)
+		return getOpponentId(game, endResult.forfeit!.playerId)
 	}
 	if (Object.hasOwn(endResult, 'playerRemoved')) {
-		return getOpponentId(game, endResult.playerRemoved.player.id)
+		return getOpponentId(game, endResult.playerRemoved!.player.id)
```

Comment:
I would prefer if you used an assert statement instead of the bang operator. This section of the code is pretty complex so when something does go wrong I want good feedback there.